server/src/main/java/org/elasticsearch/search/vectors/KnnVectorQueryBuilder.java

```diff
@@ -306,6 +307,15 @@ protected Query doToQuery(SearchExecutionContext context) throws IOException {
                 parentFilter = context.bitsetFilter(Queries.newNonNestedFilter(context.indexVersionCreated()));
             }
             if (filterQuery != null) {
+                NestedObjectMapper mapper = context.nestedLookup().getNestedMappers().get(parentPath);
+                NestedHelper nestedHelper = new NestedHelper(context.nestedLookup(), context::isFieldMapped);
+                // We treat the provided filter as a filter over PARENT documents, so if it might match nested documents
+                // we need to adjust it.
+                if (nestedHelper.mightMatchNestedDocs(filterQuery)) {
+                    // Ensure that the query only returns parent documents matching `filterQuery`
+                    filterQuery = Queries.filtered(filterQuery, Queries.not(mapper.nestedTypeFilter()));
```

Comment:
> The query could still match nested docs from another path?

I am not sure what you mean by this. There is only a single filter passed into kNN and it is the only thing we are concerned about as we are querying which child blocks match the filter.

So, the filter lifespan goes like this:
 - Determine the matching parent documents given the pre-filter
 - Join those to the child blocks
 - Use that child block pre-filter when exploring the HNSW graph because vectors are all in the child blocks.

We use the parent bit set to return back to the parent level space given the matched children.

> Since we want to restrict to parent documents can we reuse the parentFilter that is created above and turn it into a bitset filter when creating the final query?

By "final query" you mean the "final prefilter" or the "final kNN Query"?

Reply:
The contract of the `ToChildBlockJoinQuery` is that the query (the `filterQuery` here) must match documents that are present in the bitset filter (parentFilter in this context). In order to do so this change adds:
`Queries.filtered(filterQuery, Queries.not(mapper.nestedTypeFilter()));`
I don't think this is enough since the negation part only ensures that nested document from the current path are not included. If you have another nested field in your mapping `Queries.not(mapper.nestedTypeFilter())` will not filter them. That's why I think we need a filter that identifies parent document instead of excluding nested documents from the current path. 

> By "final query" you mean the "final prefilter" or the "final kNN Query"?

I mean something like:
```
if (parentPath != null) {
            NestedObjectMapper originalObjectMapper = context.nestedScope().getObjectMapper();
            final Query parentFilter;
            if (originalObjectMapper != null) {
                try {
                    // we are in a nested context, to get the parent filter we need to go up one level
                    context.nestedScope().previousLevel();
                    NestedObjectMapper objectMapper = context.nestedScope().getObjectMapper();
                    parentFilter = objectMapper == null
                            ? Queries.newNonNestedFilter(context.indexVersionCreated())
                            : objectMapper.nestedTypeFilter();
                } finally {
                    context.nestedScope().nextLevel(originalObjectMapper);
                }
            } else {
                // we are NOT in a nested context, coming from the top level knn search
                parentFilter = Queries.newNonNestedFilter(context.indexVersionCreated());
            }
            final BitSetProducer parentBitSet = context.bitsetFilter(parentFilter);
            if (filterQuery != null) {
                // We treat the provided filter as a filter over PARENT documents, so if it might match nested documents
                // we need to adjust it.
                if (nestedHelper.mightMatchNestedDocs(filterQuery)) {
                    // Ensure that the query only returns parent documents matching `filterQuery`
                    filterQuery = Queries.filtered(filterQuery, parentFilter);
                }
                // Now join the filterQuery & parentFilter to provide the matching blocks of children
                filterQuery = new ToChildBlockJoinQuery(filterQuery, parentBitSet);
            }
            return vectorFieldType.createKnnQuery(queryVector, adjustedNumCands, filterQuery, vectorSimilarity, parentBitSet);
        }
```

Reply:
@jimczi I understand what you are saying. My mistake, I didn't understand "another path", but indeed it is true, another NESTED field could cause issues. Let me fix this.

Reply:
I fixed it similarly to how you suggested and updated my yaml test to cover the curious case. 

The yaml test fails with my previous exclusion logic, but now passes with your suggestion. Thanks!